text/0000-is-aligned-intrinsic.md

```diff
@@ -0,0 +1,287 @@
+
+The standard library (and most likely many crates) use code like
+
+```rust
+let is_aligned = (ptr as usize) & ((1 << (align - 1)) - 1) == 0;
+let is_2_word_aligned = ((ptr as usize + index) & (usize_bytes - 1)) == 0;
+let is_t_aligned = ((ptr as usize) % std::mem::align_of::<T>()) == 0;
+```
+
+to check whether a pointer is aligned in order to perform optimizations like
+reading multiple bytes at once. Not only is this code which is easy to get
+wrong, and which is hard to read (and thus increasing the chance of future breakage)
+but it also makes it impossible for `miri` to evaluate such statements. This
+means that `miri` cannot do utf8-checking, since that code contains such
+optimizations. Without utf8-checking, Rustc's future const evaluation would not
+be able to convert a `[u8]` into a `str`.
+
+# Detailed design
+[design]: #detailed-design
+
+## supporting intrinsic
+
+Add a new intrinsic
+
+```rust
+fn alignto(ptr: *const (), align: usize) -> usize;
+```
+
+which takes an arbitrary pointer it never reads from and a desired alignment
+and returns the number of bytes that the pointer needs to be offset in order
+to make it aligned to the desired alignment. It is perfectly valid for an
+implementation to always yield `usize::max_value()` to signal that the pointer
+cannot be aligned. Since the caller needs to check whether the returned offset
+would be in-bounds of the allocation that the pointer points into, returning
+`usize::max_value()` will never be in-bounds of the allocation and therefor
+the caller cannot act upon the returned offset.
+
+Most implementations will expand this intrinsic to
+
+```rust
+fn alignto(ptr: *const (), align: usize) -> usize {
+    align - (ptr as usize % align)
+}
+```
+
+## standard library functions
+
+Add a new method `alignto` to `*const T` and `*mut T`, which forwards to the
+`alignto`
+
+Add two new functions `alignto` and `alignto_mut` to `core::mem` and `std::mem`
+with the following signature:
+
+```rust
+unsafe fn alignto<T, U>(&[U]) -> (&[U], &[T], &[U]) { /**/ }
+unsafe fn alignto_mut<T, U>(&mut [U]) -> (&mut [U], &mut [T], &mut [U]) { /**/ }
+```
+
+Calls to `alignto` are expanded to
+
+```rust
+unsafe fn alignto<T, U>(slice: &[U]) -> (&[U], &[T], &[U]) {
+    if size_of::<T>() % size_of::<U>() == 0 {
+        let align = core::mem::align_of::<T>();
+        let size = core::mem::size_of::<T>();
+        let source_size = core::mem::size_of::<U>();
+        // number of bytes that need to be skipped until the pointer is aligned
+        let offset = core::intrinsics::alignto(slice.as_ptr(), align);
+        // if `align_of::<T>() <= align_of::<U>()`, or if pointer is accidentally aligned, then `offset == 0`
+        //
+        // due to `size_of::<T>() % size_of::<U>() == 0`,
+        // the fact that `size_of::<U>() > align_of::<U>()`,
+        // and the fact that `align_of::<T>() > align_of::<U>()` if `offset != 0` we know
+        // that `offset % source_size == 0`
```

Comment:
True, but in that case we get a value way bigger than `slice.len()` and thus result in the case where `head == slice` and `mid` and `tail` are empty. Which is the correct behaviour.

```rust
let head_count = offset / source_size;
let split_position = core::cmp::max(slice.len(), head_count);
let (head, tail) = slice.split_at(split_position);
```

But I just noticed that for zsts this function panics due to division by zero. Not sure what to do about that.